apis/kueue/v1beta1/clusterqueue_types.go

```diff
@@ -0,0 +1,228 @@
+)
+
+// ClusterQueueSpec defines the desired state of ClusterQueue
+type ClusterQueueSpec struct {
+	// resourceGroups describes groups of resources.
+	// Each resource group defines the list of resources and a list of flavors
+	// that provide quotas for these resources.
+	// resourceGroups can be up to 16.
+	// +listType=atomic
+	// +kubebuilder:validation:MaxItems=16
+	ResourceGroups []ResourceGroup `json:"resourceGroups,omitempty"`
+
+	// cohort that this ClusterQueue belongs to. CQs that belong to the
+	// same cohort can borrow unused resources from each other.
+	//
+	// A CQ can be a member of a single borrowing cohort. A workload submitted
+	// to a queue referencing this CQ can borrow quotas from any CQ in the
+	// cohort. Only quotas listed in the CQ can be borrowed.
+	// If empty, this ClusterQueue cannot borrow from any other ClusterQueue and
+	// vice versa.
+	//
+	// Validation of a cohort name is equivalent to label keys. These are just
+	// names to link CQs together, and they are meaningless otherwise.
+	Cohort string `json:"cohort,omitempty"`
+
+	// QueueingStrategy indicates the queueing strategy of the workloads
+	// across the queues in this ClusterQueue. This field is immutable.
+	// Current Supported Strategies:
+	//
+	// - StrictFIFO: workloads are ordered strictly by creation time.
+	// Older workloads that can't be admitted will block admitting newer
+	// workloads even if they fit available quota.
+	// - BestEffortFIFO：workloads are ordered by creation time,
+	// however older workloads that can't be admitted will not block
+	// admitting newer workloads that fit existing quota.
+	//
+	// +kubebuilder:default=BestEffortFIFO
+	// +kubebuilder:validation:Enum=StrictFIFO;BestEffortFIFO
+	QueueingStrategy QueueingStrategy `json:"queueingStrategy,omitempty"`
+
+	// namespaceSelector defines which namespaces are allowed to submit workloads to
+	// this clusterQueue. Beyond this basic support for policy, an policy agent like
+	// Gatekeeper should be used to enforce more advanced policies.
+	// Defaults to null which is a nothing selector (no namespaces eligible).
+	// If set to an empty selector `{}`, then all namespaces are eligible.
+	NamespaceSelector *metav1.LabelSelector `json:"namespaceSelector,omitempty"`
+}
+
+type QueueingStrategy string
+
+const (
+	// StrictFIFO means that workloads are ordered strictly by creation time.
+	// Older workloads that can't be admitted will block admitting newer
+	// workloads even if they fit available quota.
+	StrictFIFO QueueingStrategy = "StrictFIFO"
+
+	// BestEffortFIFO means that workloads are ordered by creation time,
+	// however older workloads that can't be admitted will not block
+	// admitting newer workloads that fit existing quota.
+	BestEffortFIFO QueueingStrategy = "BestEffortFIFO"
+)
+
+type ResourceGroup struct {
+	// resources is the list of resources covered by the flavors in this group.
+	Resources []corev1.ResourceName `json:"resources"`
```

Comment:
The field has the same name as the one in FlavorQuota but a different type. Maybe one of them should have a different name? It looks a bit weird in yaml.

Reply:
I think that's ok. They are in different contexts.

Reply:
Changed